File: engines/ts/src/packaging.ts
/**
 * Harness packaging surface (M7): inspect / test / pack helpers.
 */
import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { basename, join, resolve } from "node:path";
import {
  loadAgentDir,
  loadDeployment,
  listFixtureDirs,
  readJsonl,
} from "./load.js";
import { validateAgent } from "./validate.js";
import { AdapterError, resolveToolsFromDeployment } from "./adapters/registry.js";
import { FaustRuntime, eventsToJsonl } from "./runtime.js";
import { parseRecordedModelLine } from "./adapters/recorded.js";
import { createGreenhouseTools, createCodingTools, createSpawnTool, codingWorldFromAgent } from "./tools/world.js";
import type { AgentIR, Deployment, HarnessPatch, ModelProposal, RecordedToolCall } from "./types.js";
import { canonicalJson } from "./canonical.js";
import {
  ConnectorError,
  resolveHarness,
  resolvedHarnessHash,
} from "./connectors/resolve.js";
import { signBundle, loadSignKeyFromPath } from "./bundle-signature.js";
import { applyCandidatePatch, harnessIrHash } from "./harness-patch.js";

// load.ts may not export loadFixtureAgent — inline
function loadFixtureAgentLocal(dir: string): AgentIR {
  const json = join(dir, "agent.json");
  if (existsSync(json)) return JSON.parse(readFileSync(json, "utf8")) as AgentIR;
  return loadAgentDir(dir).agent;
}

const DEPLOYMENT_CANDIDATES = [
  "deployment.fixture.yml",
  "deployment.simulation.yml",
  "deployment.openrouter-free.yml",
  "deployment.kit.yml",
  "deployment.openai.yml",
  "deployment.ollama.yml",
];

/** Local real-I/O deployments — never auto-selected by test/replay/default discovery. */
const LOCAL_DEPLOYMENT_PREFIX = "deployment.local";

export function isLocalOnlyDeploymentFile(name: string): boolean {
  return name.startsWith(LOCAL_DEPLOYMENT_PREFIX);
}

export function listHarnessDeployments(harnessDir: string): string[] {
  return DEPLOYMENT_CANDIDATES.filter((n) => existsSync(join(harnessDir, n))).map((n) =>
    join(harnessDir, n),
  );
}

export function pickTestDeployment(harnessDir: string, explicit?: string): string | undefined {
  if (explicit) return resolve(explicit);
  for (const n of ["deployment.fixture.yml", "deployment.simulation.yml"]) {
    const p = join(harnessDir, n);
    if (existsSync(p)) return p;
  }
  return listHarnessDeployments(harnessDir).find((p) => {
    const base = p.split(/[/\\]/).pop() ?? "";
    return !isLocalOnlyDeploymentFile(base);
  });
}

export type InspectReport = {
  harness: string;
  path: string;
  name: string;
  spec: string;
  tools: string[];
  permissions_tools?: string[];
  sequences?: string[];
  mutable?: string[];
  spawn?: AgentIR["spawn"];
  deployments: { file: string; platform?: string; transport?: string; binding_count: number }[];
  smoke: { model: boolean; expected: boolean };
  binding_coverage?: {
    deployment: string;
    missing: string[];
    ok: boolean;
    error?: string;
  };
  resolution?: {
    connectors_file: boolean;
    connector_count: number;
    kinds: string[];
    selected_count: number;
    lock_count: number;
    resolved_sha256: string;
    ok: boolean;
    error?: string;
    bundle_format?: string;
    embedded?: boolean;
  };
};

export function inspectHarness(
  harnessDir: string,
  opts: {
    embeddedResolved?: import("./types.js").ResolvedHarnessIR;
    bundleFormat?: string;
  } = {},
): InspectReport {
  const dir = resolve(harnessDir);
  const { agent: sourceAgent } = loadAgentDir(dir);
  const connectorsPresent =
    existsSync(join(dir, "connectors.yml")) ||
    existsSync(join(dir, "connectors.yaml")) ||
    existsSync(join(dir, "connectors.json"));
  let agent = sourceAgent;
  let resolution: InspectReport["resolution"];
  let resolvedIr: import("./types.js").ResolvedHarnessIR | undefined;
  try {
    resolvedIr = opts.embeddedResolved ?? resolveHarness(dir);
    agent = resolvedIr.agent;
    resolution = {
      connectors_file: connectorsPresent || Boolean(opts.embeddedResolved),
      connector_count: resolvedIr.resolution.connectors.length,
      kinds: Array.from(
        new Set(resolvedIr.resolution.connectors.map((c) => c.kind)),
      ).sort(),
      selected_count: resolvedIr.resolution.selected.length,
      lock_count: resolvedIr.resolution.lock.length,
      resolved_sha256: resolvedHarnessHash(resolvedIr),
      ok: true,
      ...(opts.bundleFormat ? { bundle_format: opts.bundleFormat } : {}),
      ...(opts.embeddedResolved ? { embedded: true } : {}),
    };
  } catch (e) {
    resolution = {
      connectors_file: connectorsPresent,
      connector_count: 0,
      kinds: [],
      selected_count: 0,
      lock_count: 0,
      resolved_sha256: "",
      ok: false,
      error: e instanceof ConnectorError || e instanceof Error ? e.message : String(e),
      ...(opts.bundleFormat ? { bundle_format: opts.bundleFormat } : {}),
    };
  }
  const auto = listHarnessDeployments(dir);
  const localExtras = readdirSync(dir)
    .filter((n) => isLocalOnlyDeploymentFile(n) && /\.ya?ml$/i.test(n))
    .map((n) => join(dir, n));
  const deployments = [...auto, ...localExtras].map((p) => {
    const d = loadDeployment(p) as Deployment;
    return {
      file: basename(p),
      platform: d.platform,
      transport: d.model?.transport,
      binding_count: Object.keys(d.bindings ?? {}).length,
    };
  });

  const report: InspectReport = {
    harness: basename(dir),
    path: dir,
    name: agent.name,
    spec: agent.spec,
    tools: agent.tools.map((t) => t.id),
    permissions_tools: agent.permissions?.tools,
    sequences: agent.counterbalance?.sequences?.map((s) => s.id ?? s.action),
    mutable: agent.mutable,
    spawn: agent.spawn,
    deployments,
    smoke: {
      model: existsSync(join(dir, "smoke.model.jsonl")),
      expected: existsSync(join(dir, "smoke.expected.jsonl")),
    },
    resolution,
  };

  const depPath = pickTestDeployment(dir);
  if (depPath) {
    try {
      const dep = loadDeployment(depPath) as Deployment;
      resolveToolsFromDeployment(agent, dep, {
        harnessDir: dir,
        resolved: resolvedIr,
      });
      report.binding_coverage = {
        deployment: basename(depPath),
        missing: [],
        ok: true,
      };
    } catch (e) {
      const missing: string[] = [];
      if (e instanceof AdapterError && e.code === "binding_missing") {
        const m = /tool '([^']+)'/.exec(e.message);
        if (m) missing.push(m[1]!);
      }
      report.binding_coverage = {
        deployment: basename(depPath),
        missing,
        ok: false,
        error: e instanceof Error ? e.message : String(e),
      };
    }
  }

  return report;
}

function buildTools(agent: AgentIR) {
  return {
    ...createGreenhouseTools({
      temperature_decidegrees: Number(agent.state.temperature_decidegrees ?? 250),
      fan_percent: Number(agent.state.fan_percent ?? 0),
      sensor_healthy: Number(agent.state.sensor_healthy ?? 1),
    }),
    ...createCodingTools(codingWorldFromAgent(agent)),
    ...createSpawnTool(agent.permissions?.tools ?? []),
  };
}

async function replayFixtureDir(
  dir: string,
): Promise<{ ok: boolean; name: string }> {
  const name = basename(dir);
  const agent = loadFixtureAgentLocal(dir);
  const proposals: ModelProposal[] = readJsonl(join(dir, "model.jsonl")).map(parseRecordedModelLine);
  const toolsQueue = existsSync(join(dir, "tools.jsonl"))
    ? (readJsonl(join(dir, "tools.jsonl")) as RecordedToolCall[])
    : [];
  let pi = 0;
  const runtime = new FaustRuntime({
    agent,
    propose: async () => (pi >= proposals.length ? { type: "stop" } : proposals[pi++]!),
    tools: buildTools(agent),
    recordedToolResults: toolsQueue.length ? toolsQueue : undefined,
    allowJudge: false,
  });
  await runtime.runLoop();
  const actual = eventsToJsonl(runtime.events);
  const expected = readFileSync(join(dir, "expected.jsonl"), "utf8").replace(/\r\n/g, "\n");
  return { ok: actual === expected, name };
}

export type TestHarnessResult = {
  ok: boolean;
  validate_ok: boolean;
  bindings_ok: boolean;
  smoke_ok: boolean | null;
  fixtures_ok: boolean | null;
  errors: string[];
  details: string[];
};

async function runSmoke(
  agent: AgentIR,
  deployment: Deployment,
  harnessDir: string,
  resolved?: import("./types.js").ResolvedHarnessIR,
): Promise<{ ok: boolean; detail: string }> {
  const modelPath = join(harnessDir, "smoke.model.jsonl");
  const expectedPath = join(harnessDir, "smoke.expected.jsonl");
  if (!existsSync(modelPath)) {
    return { ok: true, detail: "no smoke.model.jsonl (skipped)" };
  }
  const tools = resolveToolsFromDeployment(agent, deployment, {
    harnessDir,
    resolved,
  });
  const proposals: ModelProposal[] = readJsonl(modelPath).map(parseRecordedModelLine);
  let pi = 0;
  const rt = new FaustRuntime({
    agent,
    propose: async () => (pi >= proposals.length ? { type: "stop" } : proposals[pi++]!),
    tools,
  });
  await rt.runLoop(32);
  const actual = eventsToJsonl(rt.events);
  if (!existsSync(expectedPath)) {
    return { ok: false, detail: "smoke.model.jsonl present but smoke.expected.jsonl missing" };
  }
  const expected = readFileSync(expectedPath, "utf8").replace(/\r\n/g, "\n");
  if (actual !== expected) {
    return { ok: false, detail: "smoke event log ≠ smoke.expected.jsonl" };
  }
  return { ok: true, detail: "smoke OK" };
}

function fixturePrefixesForHarness(harnessName: string): string[] {
  if (harnessName === "coding-counterbalance") {
    return ["cb-coding-", "cb-write-", "cb-stale-", "cb-completion-", "cb-user-", "cb-harness-patch-"];
  }
  if (harnessName === "mutable-skills") return ["cb-harness-patch-"];
  if (harnessName === "support-bot") return ["cb-support-"];
  if (harnessName === "coding") return ["code-", "spawn-"];
  if (harnessName === "greenhouse") {
    return ["allow-", "deny-", "verify-", "budget-", "predicate-", "input-"];
  }
  return [];
}

export async function testHarness(
  harnessDir: string,
  opts: {
    deployment?: string;
    fixturesRoot?: string;
    skipFixtures?: boolean;
    embeddedResolved?: import("./types.js").ResolvedHarnessIR;
    bundleFormat?: string;
  } = {},
): Promise<TestHarnessResult> {
  const dir = resolve(harnessDir);
  const errors: string[] = [];
  const details: string[] = [];

  let resolved;
  try {
    resolved = opts.embeddedResolved ?? resolveHarness(dir);
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    return {
      ok: false,
      validate_ok: false,
      bindings_ok: false,
      smoke_ok: null,
      fixtures_ok: null,
      errors: [message],
      details: [],
    };
  }
  if (opts.bundleFormat) details.push(`bundle ${opts.bundleFormat}`);
  if (opts.embeddedResolved) {
    details.push(
      `embedded resolved OK (sha256=${resolvedHarnessHash(resolved)}, ${resolved.resolution.connectors.length} connectors)`,
    );
  } else {
    details.push(`resolve OK (${resolved.resolution.connectors.length} connectors)`);
  }
  const v = validateAgent(resolved.agent);
  const validate_ok = v.ok;
  if (!v.ok) {
    errors.push(...v.errors);
    return {
      ok: false,
      validate_ok: false,
      bindings_ok: false,
      smoke_ok: null,
      fixtures_ok: null,
      errors,
      details,
    };
  }
  details.push("validate OK");
  for (const w of v.warnings) details.push(`warn: ${w}`);

  const agent = resolved.agent;
  const depPath = pickTestDeployment(dir, opts.deployment);
  if (!depPath) {
    errors.push("no deployment found for harness test (need fixture/simulation or --deployment)");
    return {
      ok: false,
      validate_ok,
      bindings_ok: false,
      smoke_ok: null,
      fixtures_ok: null,
      errors,
      details,
    };
  }
  const deployment = loadDeployment(depPath) as Deployment;
  let bindings_ok = true;
  try {
    resolveToolsFromDeployment(agent, deployment, {
      harnessDir: dir,
      resolved,
    });
    details.push(`bindings OK (${basename(depPath)})`);
  } catch (e) {
    bindings_ok = false;
    errors.push(e instanceof Error ? e.message : String(e));
  }

  let smoke_ok: boolean | null = null;
  if (bindings_ok) {
    try {
      const smoke = await runSmoke(agent, deployment, dir, resolved);
      smoke_ok = smoke.ok;
      details.push(smoke.detail);
      if (!smoke.ok) errors.push(smoke.detail);
    } catch (e) {
      smoke_ok = false;
      errors.push(e instanceof Error ? e.message : String(e));
    }
  }

  let fixtures_ok: boolean | null = null;
  if (!opts.skipFixtures) {
    const prefixes = fixturePrefixesForHarness(basename(dir));
    const fixturesRoot =
      opts.fixturesRoot ??
      (existsSync(join(dir, "../../conformance/fixtures"))
        ? resolve(dir, "../../conformance/fixtures")
        : resolve(dir, "../../../conformance/fixtures"));
    if (prefixes.length && existsSync(fixturesRoot)) {
      const dirs = listFixtureDirs(fixturesRoot).filter((d) =>
        prefixes.some((p) => basename(d).startsWith(p)),
      );
      if (dirs.length) {
        fixtures_ok = true;
        for (const d of dirs) {
          const r = await replayFixtureDir(d);
          details.push(`${r.ok ? "PASS" : "FAIL"} fixture ${r.name}`);
          if (!r.ok) {
            fixtures_ok = false;
            errors.push(`fixture ${r.name} mismatch`);
          }
        }
      }
    }
  }

  const ok =
    validate_ok &&
    bindings_ok &&
    (smoke_ok === null || smoke_ok) &&
    (fixtures_ok === null || fixtures_ok);

  return { ok, validate_ok, bindings_ok, smoke_ok, fixtures_ok, errors, details };
}

const PACK_INCLUDE = [
  "agent.yml",
  "agent.json",
  "README.md",
  "smoke.model.jsonl",
  "smoke.expected.jsonl",
  ...DEPLOYMENT_CANDIDATES,
];

const CONNECTOR_MANIFEST_NAMES = ["connectors.yml", "connectors.yaml", "connectors.json"];

function collectPackSourceFiles(dir: string): { path: string; content: string }[] {
  const files: { path: string; content: string }[] = [];
  const seen = new Set<string>();
  for (const n of PACK_INCLUDE) {
    const p = join(dir, n);
    if (existsSync(p) && statSync(p).isFile()) {
      files.push({ path: n, content: readFileSync(p, "utf8") });
      seen.add(n);
    }
  }
  for (const ent of readdirSync(dir).sort()) {
    if (/^deployment\..+\.ya?ml$/i.test(ent) && !seen.has(ent)) {
      files.push({ path: ent, content: readFileSync(join(dir, ent), "utf8") });
    }
  }
  return files;
}

/**
 * Pack harness into a portable `.fausth.json` bundle (git/archives before registry).
 * Manifest-less harnesses → byte-identical v0.1.
 * Connector harnesses without mcp → v0.2.
 * MCP connector harnesses → v0.3.
 * Optional Ed25519 signing via `signKey` path or FAUSTH_SIGN_KEY (opt-in; default unsigned).
 */
export function packHarness(
  harnessDir: string,
  outPath?: string,
  opts: { signKey?: string } = {},
): { out: string; files: string[]; format: string; signed: boolean } {
  const dir = resolve(harnessDir);
  const name = basename(dir);
  const hasConnectors = CONNECTOR_MANIFEST_NAMES.some(
    (n) => existsSync(join(dir, n)) && statSync(join(dir, n)).isFile(),
  );

  const files = collectPackSourceFiles(dir);
  // Optional MCP recorded fixture for Track A packs
  for (const extra of ["mcp.recorded.jsonl", "module.recorded.jsonl"]) {
    const p = join(dir, extra);
    if (existsSync(p) && statSync(p).isFile() && !files.some((f) => f.path === extra)) {
      files.push({ path: extra, content: readFileSync(p, "utf8") });
    }
  }

  let bundle: Record<string, unknown>;
  if (hasConnectors) {
    const resolved = resolveHarness(dir);
    const hasMcp = resolved.resolution.lock.some((l) => l.kind === "mcp");
    const hasModule = resolved.resolution.lock.some((l) => l.kind === "module");
    for (const n of CONNECTOR_MANIFEST_NAMES) {
      const p = join(dir, n);
      if (existsSync(p) && statSync(p).isFile()) {
        files.push({ path: n, content: readFileSync(p, "utf8") });
      }
    }
    for (const lock of resolved.resolution.lock) {
      if ((lock.kind !== "file" && lock.kind !== "mcp" && lock.kind !== "module") || !lock.path) continue;
      const p = join(dir, lock.path);
      if (!existsSync(p) || !statSync(p).isFile()) {
        throw new ConnectorError(
          "connector_import_not_found",
          `pack: lock path missing on disk: ${lock.path}`,
        );
      }
      if (!files.some((f) => f.path === lock.path)) {
        files.push({ path: lock.path, content: readFileSync(p, "utf8") });
      }
    }
    files.sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0));
    bundle = {
      format: hasMcp || hasModule ? "fausth-harness-bundle/v0.3" : "fausth-harness-bundle/v0.2",
      name,
      files: Object.fromEntries(files.map((f) => [f.path, f.content])),
      resolved,
      resolved_sha256: resolvedHarnessHash(resolved),
    };
  } else {
    files.sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0));
    bundle = {
      format: "fausth-harness-bundle/v0.1",
      name,
      files: Object.fromEntries(files.map((f) => [f.path, f.content])),
    };
  }

  const signKey = opts.signKey ?? process.env.FAUSTH_SIGN_KEY;
  let signed = false;
  if (signKey) {
    bundle = signBundle(bundle, loadSignKeyFromPath(signKey));
    signed = true;
  }

  let out: string;
  if (outPath && outPath.endsWith(".fausth.json")) {
    out = resolve(outPath);
    mkdirSync(join(out, ".."), { recursive: true });
  } else {
    const outDir = outPath ? resolve(outPath) : join(dir, "dist");
    mkdirSync(outDir, { recursive: true });
    out = join(outDir, `${name}.fausth.json`);
  }
  writeFileSync(out, canonicalJson(bundle) + "\n", "utf8");
  return {
    out,
    files: files.map((f) => f.path),
    format: String(bundle.format),
    signed,
  };
}

export type SelectResult = {
  ok: boolean;
  harness_hash_before: string;
  harness_hash_after: string;
  fixtures_ok: boolean;
  details: string[];
  errors: string[];
  patched_agent?: AgentIR;
};

/**
 * Selection gate: apply a candidate patch, then re-run matching Track A fixtures.
 * Any golden failure rejects the mutation. Reproduction is `fausth pack` of the patched harness.
 */
export async function selectHarness(
  harnessDir: string,
  patch: HarnessPatch,
  opts: { fixturesRoot?: string; skipFixtures?: boolean } = {},
): Promise<SelectResult> {
  const dir = resolve(harnessDir);
  const { agent } = loadAgentDir(dir);
  const before = harnessIrHash(agent);
  const details: string[] = [];
  const errors: string[] = [];

  let patched: AgentIR;
  try {
    patched = applyCandidatePatch(agent, patch);
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    return {
      ok: false,
      harness_hash_before: before,
      harness_hash_after: before,
      fixtures_ok: false,
      details,
      errors: [msg],
    };
  }
  const after = harnessIrHash(patched);
  details.push(`patch applied ${before.slice(0, 12)}… → ${after.slice(0, 12)}…`);

  if (opts.skipFixtures) {
    return {
      ok: true,
      harness_hash_before: before,
      harness_hash_after: after,
      fixtures_ok: true,
      details: [...details, "fixtures skipped"],
      errors,
      patched_agent: patched,
    };
  }

  const fixturesRoot =
    opts.fixturesRoot ?? join(dir, "../../conformance/fixtures");
  let prefixes = fixturePrefixesForHarness(basename(dir));
  if (!prefixes.length && agent.mutable?.length) prefixes = ["cb-harness-patch-"];

  let fixturesOk = true;
  if (!prefixes.length) {
    details.push("no fixture prefixes; selection accepted on patch validation alone");
  } else if (!existsSync(fixturesRoot)) {
    errors.push(`fixtures root missing: ${fixturesRoot}`);
    fixturesOk = false;
  } else {
    const dirs = listFixtureDirs(fixturesRoot).filter((d) =>
      prefixes.some((p) => basename(d).startsWith(p)),
    );
    for (const fd of dirs) {
      const r = await replayFixtureDir(fd);
      if (!r.ok) {
        fixturesOk = false;
        errors.push(`${r.name}: golden mismatch`);
      } else {
        details.push(`${r.name}: OK`);
      }
    }
  }

  return {
    ok: fixturesOk && errors.length === 0,
    harness_hash_before: before,
    harness_hash_after: after,
    fixtures_ok: fixturesOk,
    details,
    errors,
    patched_agent: patched,
  };
}

File: engines/py/fausth/packaging.py
"""Harness packaging helpers (M7) — inspect / test / pack."""
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from .canonical import canonical_json
from .registry import AdapterError, load_agent_dir, load_yaml, resolve_tools_from_deployment
from .runtime import FaustRuntime, events_to_jsonl, replay_fixture
from .connectors import ConnectorError, resolve_harness, resolved_harness_hash
from .bundle_signature import load_sign_key_from_path, sign_bundle

DEPLOYMENT_CANDIDATES = [
    "deployment.fixture.yml",
    "deployment.simulation.yml",
    "deployment.openrouter-free.yml",
    "deployment.kit.yml",
    "deployment.openai.yml",
    "deployment.ollama.yml",
]

LOCAL_DEPLOYMENT_PREFIX = "deployment.local"

PACK_INCLUDE = [
    "agent.yml",
    "agent.json",
    "README.md",
    "smoke.model.jsonl",
    "smoke.expected.jsonl",
    *DEPLOYMENT_CANDIDATES,
]

CONNECTOR_MANIFEST_NAMES = ("connectors.yml", "connectors.yaml", "connectors.json")


def is_local_only_deployment_file(name: str) -> bool:
    return name.startswith(LOCAL_DEPLOYMENT_PREFIX)


def list_deployments(harness_dir: Path) -> list[Path]:
    return [harness_dir / n for n in DEPLOYMENT_CANDIDATES if (harness_dir / n).is_file()]


def pick_test_deployment(harness_dir: Path, explicit: str | None = None) -> Path | None:
    if explicit:
        return Path(explicit).resolve()
    for n in ("deployment.fixture.yml", "deployment.simulation.yml"):
        p = harness_dir / n
        if p.is_file():
            return p
    for p in list_deployments(harness_dir):
        if not is_local_only_deployment_file(p.name):
            return p
    return None


def inspect_harness(
    harness_dir: str,
    *,
    embedded_resolved: dict[str, Any] | None = None,
    bundle_format: str | None = None,
) -> dict[str, Any]:
    d = Path(harness_dir).resolve()
    source_agent = load_agent_dir(str(d))
    agent = source_agent
    connectors_present = any(
        (d / name).is_file() for name in ("connectors.yml", "connectors.yaml", "connectors.json")
    )
    resolved: dict[str, Any] | None = None
    try:
        resolved = embedded_resolved if embedded_resolved is not None else resolve_harness(d)
        agent = resolved["agent"]
        resolution = {
            "connectors_file": connectors_present or embedded_resolved is not None,
            "connector_count": len(resolved["resolution"]["connectors"]),
            "kinds": sorted(
                {c["kind"] for c in resolved["resolution"]["connectors"]}
            ),
            "selected_count": len(resolved["resolution"]["selected"]),
            "lock_count": len(resolved["resolution"]["lock"]),
            "resolved_sha256": resolved_harness_hash(resolved),
            "ok": True,
        }
        if bundle_format:
            resolution["bundle_format"] = bundle_format
        if embedded_resolved is not None:
            resolution["embedded"] = True
    except ConnectorError as e:
        resolution = {
            "connectors_file": connectors_present,
            "connector_count": 0,
            "kinds": [],
            "selected_count": 0,
            "lock_count": 0,
            "resolved_sha256": "",
            "ok": False,
            "error": str(e),
        }
        if bundle_format:
            resolution["bundle_format"] = bundle_format
        resolved = None
    deployments = []
    seen: set[str] = set()
    for p in list_deployments(d):
        seen.add(p.name)
        dep = load_yaml(str(p))
        deployments.append(
            {
                "file": p.name,
                "platform": dep.get("platform"),
                "transport": (dep.get("model") or {}).get("transport"),
                "binding_count": len(dep.get("bindings") or {}),
            }
        )
    for p in sorted(d.glob("deployment.local*.y*ml")):
        if p.name in seen or not p.is_file():
            continue
        dep = load_yaml(str(p))
        deployments.append(
            {
                "file": p.name,
                "platform": dep.get("platform"),
                "transport": (dep.get("model") or {}).get("transport"),
                "binding_count": len(dep.get("bindings") or {}),
            }
        )
    report: dict[str, Any] = {
        "harness": d.name,
        "path": str(d),
        "name": agent.get("name"),
        "spec": agent.get("spec"),
        "tools": [t["id"] for t in agent.get("tools") or []],
        "permissions_tools": (agent.get("permissions") or {}).get("tools"),
        "sequences": [
            s.get("id") or s.get("action")
            for s in (agent.get("counterbalance") or {}).get("sequences") or []
        ],
        "spawn": agent.get("spawn"),
        "deployments": deployments,
        "smoke": {
            "model": (d / "smoke.model.jsonl").is_file(),
            "expected": (d / "smoke.expected.jsonl").is_file(),
        },
        "resolution": resolution,
    }
    dep_path = pick_test_deployment(d)
    if dep_path:
        try:
            resolve_tools_from_deployment(
                agent,
                load_yaml(str(dep_path)),
                harness_dir=str(d),
                resolved=resolved,
            )
            report["binding_coverage"] = {
                "deployment": dep_path.name,
                "missing": [],
                "ok": True,
            }
        except AdapterError as e:
            report["binding_coverage"] = {
                "deployment": dep_path.name,
                "missing": [],
                "ok": False,
                "error": str(e),
            }

    return report


def test_harness(
    harness_dir: str,
    *,
    deployment: str | None = None,
    skip_fixtures: bool = False,
    embedded_resolved: dict[str, Any] | None = None,
    bundle_format: str | None = None,
) -> dict[str, Any]:
    d = Path(harness_dir).resolve()
    errors: list[str] = []
    details: list[str] = []
    try:
        resolved = embedded_resolved if embedded_resolved is not None else resolve_harness(d)
        agent = resolved["agent"]
        if bundle_format:
            details.append(f"bundle {bundle_format}")
        if embedded_resolved is not None:
            details.append(
                f"embedded resolved OK (sha256={resolved_harness_hash(resolved)}, "
                f"{len(resolved['resolution']['connectors'])} connectors)"
            )
        else:
            details.append(
                f"resolve OK ({len(resolved['resolution']['connectors'])} connectors)"
            )
    except ConnectorError as e:
        return {
            "ok": False,
            "bindings_ok": False,
            "smoke_ok": None,
            "fixtures_ok": None,
            "errors": [str(e)],
            "details": details,
        }
    dep_path = pick_test_deployment(d, deployment)
    if not dep_path:
        return {
            "ok": False,
            "errors": ["no deployment found"],
            "details": details,
        }
    try:
        tools = resolve_tools_from_deployment(
            agent,
            load_yaml(str(dep_path)),
            harness_dir=str(d),
            resolved=resolved,
        )
        details.append(f"bindings OK ({dep_path.name})")
        bindings_ok = True
    except AdapterError as e:
        errors.append(str(e))
        bindings_ok = False
        tools = {}

    smoke_ok: bool | None = None
    model_path = d / "smoke.model.jsonl"
    expected_path = d / "smoke.expected.jsonl"
    if bindings_ok and model_path.is_file():
        proposals = [
            json.loads(line)
            for line in model_path.read_text(encoding="utf-8").splitlines()
            if line.strip()
        ]
        rt = FaustRuntime(agent, proposals, tools, None)
        rt.run_loop(32)
        actual = events_to_jsonl(rt.events)
        if not expected_path.is_file():
            smoke_ok = False
            errors.append("smoke.expected.jsonl missing")
        else:
            expected = expected_path.read_text(encoding="utf-8").replace("\r\n", "\n")
            smoke_ok = actual == expected
            details.append("smoke OK" if smoke_ok else "smoke mismatch")
            if not smoke_ok:
                errors.append("smoke event log ≠ smoke.expected.jsonl")
    elif not model_path.is_file():
        details.append("no smoke.model.jsonl (skipped)")

    fixtures_ok: bool | None = None
    if not skip_fixtures:
        root = d.parents[1] / "conformance" / "fixtures"
        prefixes: tuple[str, ...] = ()
        if d.name == "coding-counterbalance":
            prefixes = (
                "cb-coding-",
                "cb-write-",
                "cb-stale-",
                "cb-completion-",
                "cb-user-",
            )
        elif d.name == "support-bot":
            prefixes = ("cb-support-",)
        if prefixes and root.is_dir():
            fixtures_ok = True
            for child in sorted(root.iterdir()):
                if child.is_dir() and child.name.startswith(prefixes):
                    ok, _, _ = replay_fixture(child)
                    details.append(f"{'PASS' if ok else 'FAIL'} fixture {child.name}")
                    if not ok:
                        fixtures_ok = False
                        errors.append(f"fixture {child.name} mismatch")

    ok = bindings_ok and (smoke_ok is None or smoke_ok) and (fixtures_ok is None or fixtures_ok)
    return {
        "ok": ok,
        "bindings_ok": bindings_ok,
        "smoke_ok": smoke_ok,
        "fixtures_ok": fixtures_ok,
        "errors": errors,
        "details": details,
    }


def pack_harness(
    harness_dir: str,
    out: str | None = None,
    *,
    sign_key: str | None = None,
) -> dict[str, Any]:
    d = Path(harness_dir).resolve()
    files: dict[str, str] = {}
    for n in PACK_INCLUDE:
        p = d / n
        if p.is_file():
            files[n] = p.read_text(encoding="utf-8")
    for p in sorted(d.iterdir()):
        if (
            p.is_file()
            and p.name.startswith("deployment.")
            and p.suffix.lower() in (".yml", ".yaml")
            and p.name not in files
        ):
            files[p.name] = p.read_text(encoding="utf-8")

    has_connectors = any((d / name).is_file() for name in CONNECTOR_MANIFEST_NAMES)
    recorded = d / "mcp.recorded.jsonl"
    if recorded.is_file():
        files["mcp.recorded.jsonl"] = recorded.read_text(encoding="utf-8")

    if has_connectors:
        resolved = resolve_harness(d)
        has_mcp = any(lock.get("kind") == "mcp" for lock in resolved["resolution"]["lock"])
        for name in CONNECTOR_MANIFEST_NAMES:
            p = d / name
            if p.is_file():
                files[name] = p.read_text(encoding="utf-8")
        for lock in resolved["resolution"]["lock"]:
            if lock.get("kind") not in ("file", "mcp"):
                continue
            path = lock.get("path")
            if not path:
                continue
            p = d / path
            if not p.is_file():
                raise ConnectorError(
                    "connector_import_not_found",
                    f"pack: lock path missing on disk: {path}",
                )
            if path not in files:
                files[path] = p.read_text(encoding="utf-8")
        ordered = {k: files[k] for k in sorted(files.keys())}
        bundle: dict[str, Any] = {
            "format": (
                "fausth-harness-bundle/v0.3" if has_mcp else "fausth-harness-bundle/v0.2"
            ),
            "name": d.name,
            "files": ordered,
            "resolved": resolved,
            "resolved_sha256": resolved_harness_hash(resolved),
        }
    else:
        ordered = {k: files[k] for k in sorted(files.keys())}
        bundle = {
            "format": "fausth-harness-bundle/v0.1",
            "name": d.name,
            "files": ordered,
        }

    key_path = sign_key if sign_key is not None else os.environ.get("FAUSTH_SIGN_KEY")
    signed = False
    if key_path:
        bundle = sign_bundle(bundle, load_sign_key_from_path(key_path))
        signed = True

    if out and out.endswith(".fausth.json"):
        out_path = Path(out).resolve()
        out_path.parent.mkdir(parents=True, exist_ok=True)
    else:
        out_dir = Path(out).resolve() if out else d / "dist"
        out_dir.mkdir(parents=True, exist_ok=True)
        out_path = out_dir / f"{d.name}.fausth.json"
    out_path.write_text(canonical_json(bundle) + "\n", encoding="utf-8", newline="\n")
    return {
        "out": str(out_path),
        "files": sorted(ordered.keys()),
        "format": bundle["format"],
        "signed": signed,
    }
